os/pm/pm_idle.c

```diff
@@ -0,0 +1,121 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * os/pm/pm_idle.c
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <tinyara/config.h>
+#include <debug.h>
+
+#include <tinyara/board.h>
+#include <tinyara/pm/pm.h>
+
+#include <tinyara/irq.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define PM_IDLE_DOMAIN	0
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: up_idlepm
+ *
+ * Description:
+ *   Perform IDLE state power management.
+ *
+ ****************************************************************************/
+#ifdef CONFIG_PM
+static void up_idlepm(void)
+{
+	static int oldstate = PM_NORMAL;
+	int newstate;
+	irqstate_t flags;
+	int ret;
+
+	/* Decide, which power saving level can be obtained */
+	newstate = pm_checkstate(PM_IDLE_DOMAIN);
+
+	/* Check for state changes */
+	if (newstate != oldstate) {
+		flags = irqsave();
+
+		/* Then force the global state change */
+		ret = pm_changestate(PM_IDLE_DOMAIN, newstate);
+		if (ret < 0) {
+			/* The new state change failed, revert to the preceding state */
+			pm_changestate(PM_IDLE_DOMAIN, oldstate);
+		} else {
+			/* Save the new state */
+			oldstate = newstate;
+		}
+
+		/* SOC-specific power management logic */
+		ret = up_pmidle_switch_state(newstate);
+		if (ret < 0) {
+			/* Staying in same sleep state as SOC returned error */
+			up_pmidle_switch_state(newstate - 1);
```

Comment:
why this is needed, if soc fails to witch newstate, it will be in previous state only right?

Reply:
@lokeshbv if in case the Soc is transitioning to a new state >>
It has already processed certain changes & then fails in the middle of transitioning, then we cannot say it is still in old state (if soc doesn't take care of reverting them)
Hence, above call is required